README.md
# json-superstring

The native `JSON.stringify()` method is quite fast, but it lacks safety checks for things like circular references and enumerable getter properties that throw errors.  The `json-superstring` module wraps `JSON.stringify()`, and includes safety checks for these potential issues.

__Table of Contents__

* [Usage](#usage)
* [API](#api)
* [Motivation](#motivation)
* [Performance](#performance)

## Usage

Add `json-superstring` as a dependency in `package.json`:

```sh
$ npm install json-superstring -S
```

Then simply call `json-superstring`...

```js
const jsonSuperstring = require('json-superstring');

const data = {
  foo: {
    a: '1',
    b: 2
  }
};
data.foo.bar = data.foo;

console.log(jsonSuperstring(data));
// {"foo":{"a":"1","b":2,"bar":"[Circular]"}}
```

## API

### `jsonSuperstring(data [, space])`

Stringifies a value in the same manner as `JSON.stringify()`.  The key difference is circular references and getter properties that throw errors do not prevent stringification.

_Parameters_

* `data`: _(required)_ the data to stringify.

* `space`: _(optional)_ a `String` or `Number` object that's used to insert white space into the output JSON string for readability purposes.  If this is a `Number`, it indicates the number of space characters to use as white space; this number is capped at 10 (if it is greater, the value is just 10).  Values less than 1 indicate that no space should be used.  If this is a `String`, the string (or the first 10 characters of the string, if it's longer than that) is used as white space.  If this parameter is not provided (or is `null`), no white space is used.

_Returns_

A JSON string.

### `jsonSuperstring.merge(...data)`

Safely merges n-number of objects together.  The resulting object can be safely used with `JSON.stringify()`.

_Parameters_

* `...data`: _(required)_ n-number of objects to merge.

_Returns_

An object.

## Motivation

There are a number of other safe JSON stringifiers out there, and they all provide some variation of performance and features.  The goals of this project are:

1. Best in class [performance](#performance).

2. Safety checks for both circular references and getter properties that errors.

3. No side effects.  Objects passed to `json-stringify` are not modified.

4. Provide ability to specify space param, consistent with `JSON.stringify()`'s functionality.

5. Provide ability to merge multiple objects together and stringify.

### Comparison

| Name                   | White Space | Circular Check | Error Check | No Side Effects |
|------------------------|-------------|----------------|-------------|-----------------|
| __`json-superstring`__ | ✔           | ✔              | ✔           | ✔               |
| `fast-safe-stringify`  | ✕           | ✔              | ✕           | ✕               |
| `json-stringify-safe`  | ✔           | ✔[*]           | ✕           | ✔               |
| `safe-json-stringify`  | ✕           | ✔[*][**]       | ✔           | ✕               |

[*] Does not check for circular references when calling an object's `toJSON()` method.

[**] While `safe-json-stringify` performs circular reference checks, it marks all duplicate object references in a JSON object as `[Circular]` regardless of whether they are actual circular references.

## Performance

Performance metrics should always be taken with a healthy level of skepticism.  Different software performs differently as the parameters change, and as the state of the host environment changes.

Moreover, not all software provides the same functionality.  As demonstrated in the [Comparison](#comparison) table above, that is the case with `json-superstring`, and other safe JSON stringifiers.

The goal here is to show that `json-superstring`'s performance is comparable with other libraries while providing a broader set of features.

Benchmark reports are generated using [`radargun`](https://www.npmjs.com/package/radargun), and built using data from 1 million runs of each utility.

__circular.bench.js__

```
┌─────────────────────┬─────────────┬─────────────┬─────────────┐
│ NAME                │ AVG         │ MIN         │ MAX         │
╞═════════════════════╪═════════════╪═════════════╪═════════════╡
│ json-superstring    │ 7582 ns     │ 5666 ns     │ 1794442 ns  │
├─────────────────────┼─────────────┼─────────────┼─────────────┤
│ fast-safe-stringify │ 8810 ns     │ 6216 ns     │ 3692754 ns  │
├─────────────────────┼─────────────┼─────────────┼─────────────┤
│ json-stringify-safe │ 10928 ns    │ 6397 ns     │ 16823572 ns │
├─────────────────────┼─────────────┼─────────────┼─────────────┤
│ safe-json-stringify │ 8231 ns     │ 4597 ns     │ 12536094 ns │
└─────────────────────┴─────────────┴─────────────┴─────────────┘
```

__throws.bench.js__

```
┌─────────────────────┬─────────────┬─────────────┬─────────────┐
│ NAME                │ AVG         │ MIN         │ MAX         │
╞═════════════════════╪═════════════╪═════════════╪═════════════╡
│ json-superstring    │ 9626 ns     │ 6157 ns     │ 7036140 ns  │
├─────────────────────┼─────────────┼─────────────┼─────────────┤
│ safe-json-stringify │ 13183 ns    │ 8427 ns     │ 12135604 ns │
└─────────────────────┴─────────────┴─────────────┴─────────────┘
```

__merge.bench.js__

```
┌──────────────────┬────────────┬────────────┬────────────┐
│ NAME             │ AVG        │ MIN        │ MAX        │
╞══════════════════╪════════════╪════════════╪════════════╡
│ json-superstring │ 9443 ns    │ 6137 ns    │ 5428283 ns │
└──────────────────┴────────────┴────────────┴────────────┘
```


CHANGELOG.md
# Change Log

## 2.0

### 2.0.0
  * Complete rewrite.

### 2.1.0
  * Adding ability to merge multiple objects together safely.

## 1.0

### 1.0.0
  * Initial release.


test/unit/index.js
'use strict';

const { assert } = require('chai');

const jsonStringify = require('../../lib');


const date = new Date(267210300000);
const dateStr = date.toISOString();


describe('json-superstring', function() {

  describe('#stringify', function() {
    it('stringifies with spaces when option provided', function() {
      const result = jsonStringify({ foo: 'bar' }, 1);
      assert.strictEqual(result, '{\n "foo": "bar"\n}');
    });

    it('does not flag null with [Circular]', function() {
      const result = jsonStringify({ foo: null, bar: null });
      assert.strictEqual(result, '{"foo":null,"bar":null}');
    });

    it('does not flag non-objects as [Circular]', function() {
      const result = jsonStringify({ foo: 42, bar: 42 });
      assert.strictEqual(result, '{"foo":42,"bar":42}');
    });

    it('does not flag arrays as [Circular]', function() {
      const result = jsonStringify({ foo: [42], bar: [42] });
      assert.strictEqual(result, '{"foo":[42],"bar":[42]}');
    });

    it('does not flag dates as [Circular]', function() {
      const result = jsonStringify({ foo: date, bar: date });
      assert.strictEqual(result, `{"foo":"${dateStr}","bar":"${dateStr}"}`);
    });

    it('flags circular references of root with [Circular]', function() {
      const data = {
        foo: 42,
      };
      data.bar = data;
      const result = jsonStringify(data);
      assert.strictEqual(result, '{"foo":42,"bar":"[Circular]"}');
    });

    it('flags nested, circular references with [Circular]', function() {
      const data = {
        foo: { a: '1' },
      };
      data.foo.bar = data.foo;

      const result = jsonStringify(data);
      assert.strictEqual(result, '{"foo":{"a":"1","bar":"[Circular]"}}');
    });

    it('does not flag seen object references with [Circular]', function() {
      const data = {
        foo: { a: '1' },
      };
      data.bar = data.foo;

      const result = jsonStringify(data);
      assert.strictEqual(result, '{"foo":{"a":"1"},"bar":{"a":"1"}}');
    });

    it('flags seen deeply nested object references with [Circular]', function() {
      const data = {
        foo: 'hello world',
        bar: 42,
        baz: {
          qux: true,
        },
      };
      data.baz.quux = data.baz;

      const result = jsonStringify(data);
      const expected = '{"foo":"hello world","bar":42,"baz":{"qux":true,"quux":"[Circular]"}}';
      assert.strictEqual(result, expected);
    });

    it('flags getter props with [Error]', function() {
      const data = {};
      Object.defineProperty(data, 'bar', {
        enumerable: true,
        get: function () { throw new Error('Hi'); },
      });

      const result = jsonStringify(data);
      const val = '{"bar":"[Error: Hi]"}';
      assert.strictEqual(result, val);
    });
  });


  describe('#merge', function() {
    it('throws if no data provided', function() {
      assert.throws(() => {
        jsonStringify.merge();
      }, TypeError);
    });

    it('merges two objects, a has circular reference, b has error', function() {
      const a = { foo: 1 };
      a.bar = a;

      const b = {};
      Object.defineProperty(b, 'baz', {
        enumerable: true,
        get: function () { throw new Error('Hi'); },
      });

      const result = jsonStringify.merge(a, b);
      const expected = {
        foo: 1,
        bar: '[Circular]',
        baz: '[Error: Hi]',
      };

      assert.deepEqual(result, expected);

      assert.doesNotThrow(() => {
        JSON.stringify(result);
      });
    });

    it('merges an array into an object', function() {
      const a = { foo: 1 };
      a.bar = a;

      const b = [a];

      const result = jsonStringify.merge(a, b);
      const expected = {
        foo: 1,
        bar: '[Circular]',
        0: { foo: 1, bar: '[Circular]' },
      };

      assert.deepEqual(result, expected);

      assert.doesNotThrow(() => {
        JSON.stringify(result);
      });
    });

    it('merges an object into an array', function() {
      const aval = { foo: 1 };
      aval.bar = aval;
      const a = [aval];

      const b = {};
      Object.defineProperty(b, 'baz', {
        enumerable: true,
        get: function () { throw new Error('Hi'); },
      });

      const result = jsonStringify.merge(a, b);
      const expected = [{
        foo: 1,
        bar: '[Circular]',
      }];
      expected.baz = '[Error: Hi]';

      assert.deepEqual(result, expected);

      assert.doesNotThrow(() => {
        JSON.stringify(result);
      });
    });

    it('merges two arrays', function() {
      const a = [42];

      const b0 = { foo: 1 };
      b0.bar = b0;

      const b1 = {};
      Object.defineProperty(b1, 'baz', {
        enumerable: true,
        get: function () { throw new Error('Hi'); },
      });

      const b = [b0, b1];

      const result = jsonStringify.merge(a, b);
      const expected = [
        {
          foo: 1,
          bar: '[Circular]',
        },
        {
          baz: '[Error: Hi]',
        },
      ];

      assert.deepEqual(result, expected);

      assert.doesNotThrow(() => {
        JSON.stringify(result);
      });
    });
  });

});


tests/unit/index.js
'use strict';

const assert = require('chai').assert;

const jsonStringify = require('../../lib');

const date = new Date(267210300000);
const dateStr = date.toISOString();

describe('json-superstring', () => {

  it('should throw if space not string or number', () => {
    assert.throws(() => {
      jsonStringify({ foo: 'bar' }, { space: {} });
    });
  });

  it('should throw if checkProps not Boolean', () => {
    assert.throws(() => {
      jsonStringify({ foo: 'bar' }, { checkProps: 42 });
    });
  });

  it('should stringify with spaces when option provided', (done) => {
    const result = jsonStringify({ foo: 'bar' }, { space: 1 });
    assert.strictEqual(result, '{\n "foo": "bar"\n}');
    done();
  });

  it('should not flag null with [Circular]', (done) => {
    const result = jsonStringify({ foo: null, bar: null });
    assert.strictEqual(result, '{"foo":null,"bar":null}');
    done();
  });

  it('should not flag non-objects as [Circular]', (done) => {
    const result = jsonStringify({ foo: 42, bar: 42 });
    assert.strictEqual(result, '{"foo":42,"bar":42}');
    done();
  });

  it('should not flag arrays as [Circular]', (done) => {
    const result = jsonStringify({ foo: [42], bar: [42] });
    assert.strictEqual(result, '{"foo":[42],"bar":[42]}');
    done();
  });

  it('should not flag dates as [Circular]', (done) => {
    const result = jsonStringify({ foo: date, bar: date });
    assert.strictEqual(result, `{"foo":"${dateStr}","bar":"${dateStr}"}`);
    done();
  });

  it('should flag nested, circular references with [Circular]', (done) => {
    const data = {
      foo: { a: '1' }
    };
    data.foo.bar = data.foo;

    const result = jsonStringify(data);
    assert.strictEqual(result, '{"foo":{"a":"1","bar":"[Circular]"}}');
    done();
  });

  it('should flag seen object references with [Circular]', (done) => {
    const data = {
      foo: { a: '1' }
    };
    data.bar = data.foo;

    const result = jsonStringify(data);
    assert.strictEqual(result, '{"foo":{"a":"1"},"bar":"[Circular]"}');
    done();
  });

  it('should flag getter props with {throws} when checkProps=true', (done) => {
    class Test {
      constructor() {
        this.foo = { a: 1 };
        Object.defineProperty(this, 'bar', { enumerable: true });
      }
    }

    const data = {};
    Object.defineProperty(data, 'bar', {
      enumerable: true,
      get: function () { throw new Error('Hi'); }
    });

    const result = jsonStringify(data, { checkProps: true });
    const val = '{"bar":{"throws":{"name":"Error","message":"Hi"}}}';
    assert.strictEqual(result, val);
    done();
  });

});


lib/index.js
'use strict';

//
// CONSTANTS
//


const CIRCULAR = '[Circular]';
const ERR_OPEN = '[';
const ERR_SPEC = ': ';
const ERR_CLOSE = ']';


//
// SAFE COPIER
//


function copy(into, data, stack) {
  if (typeof data !== 'object' || data === null) return data;

  let result = into;

  if (result === null) {
    result = (Array.isArray(data))
      ? Array(data.length)
      : {};
  }

  for (const key in data) {
    let val;

    try {
      val = data[key];

      if (val !== null
        && typeof val !== 'undefined'
        && typeof val.toJSON === 'function'
      ) {
        val = val.toJSON();
      }
    } catch (err) {
      val = ERR_OPEN + err.name + ERR_SPEC + err.message + ERR_CLOSE;
    }

    if (typeof val === 'object') {
      if (stack.indexOf(val) > -1) {
        val = CIRCULAR;
      } else {
        stack.push(val);
        val = copy(null, val, stack);
        stack.pop();
      }
    }

    result[key] = val;
  }

  return result;
}


//
// STRINGIFY FUNCTION
//


function stringify(data, space) {
  const result = copy(null, data, [data]);
  if (!space) return JSON.stringify(result);
  return JSON.stringify(result, null, space);
}


//
// MERGE FUNCTION
//


function merge(...data) {
  if (data.length === 0) {
    throw new TypeError('At least one object is required');
  }

  let result = null;

  for (let i = 0; i < data.length; i++) {
    const d = data[i];
    result = copy(result, d, [d]);
  }

  return result;
}


stringify.merge = merge;


module.exports = stringify;


benchmark/throws.bench.js
'use strict';

const safeJsonStringify = require('safe-json-stringify');

const jsonSuperstring = require('../lib');


bench(
  [
    {
      label: 'json-superstring',
      fn: function benchJsonSuperstring() {
        const value = {
          foo: 42,
        };
        Object.defineProperty(value, 'bar', {
          enumerable: true,
          get: function () { throw new Error('Hi'); },
        });

        jsonSuperstring(value);
      },
    },
    {
      label: 'safe-json-stringify',
      fn: function benchSafeJsonStringify() {
        const value = {
          foo: 42,
        };
        Object.defineProperty(value, 'bar', {
          enumerable: true,
          get: function () { throw new Error('Hi'); },
        });

        safeJsonStringify(value);
      },
    },
  ],
  { runs: 1000000 }
);


benchmark/circular.bench.js
'use strict';

const fastSafeStringify = require('fast-safe-stringify');
const jsonStringifySafe = require('json-stringify-safe');
const safeJsonStringify = require('safe-json-stringify');

const jsonSuperstring = require('../lib');


bench(
  [
    {
      label: 'json-superstring',
      fn: function benchJsonSuperstring() {
        const value = {
          foo: {
            bar: 'baz',
            qux: 42,
            quux: {},
            corge: { grault: true },
            waldo: [],
          },
          blah: {},
        };
        value.foo.quux.quuz = value.foo;
        value.foo.garply = value.foo.corge;
        value.foo.waldo.push(value.foo);
        value.blah.huh = value.blah;

        jsonSuperstring(value);
      },
    },
    {
      label: 'fast-safe-stringify',
      fn: function benchFastSafeStringify() {
        const value = {
          foo: {
            bar: 'baz',
            qux: 42,
            quux: {},
            corge: { grault: true },
            waldo: [],
          },
          blah: {},
        };
        value.foo.quux.quuz = value.foo;
        value.foo.garply = value.foo.corge;
        value.foo.waldo.push(value.foo);
        value.blah.huh = value.blah;

        fastSafeStringify(value);
      },
    },
    {
      label: 'json-stringify-safe',
      fn: function benchJsonStringifySafe() {
        const value = {
          foo: {
            bar: 'baz',
            qux: 42,
            quux: {},
            corge: { grault: true },
            waldo: [],
          },
          blah: {},
        };
        value.foo.quux.quuz = value.foo;
        value.foo.garply = value.foo.corge;
        value.foo.waldo.push(value.foo);
        value.blah.huh = value.blah;

        jsonStringifySafe(value);
      },
    },
    {
      label: 'safe-json-stringify',
      fn: function benchSafeJsonStringify() {
        const value = {
          foo: {
            bar: 'baz',
            qux: 42,
            quux: {},
            corge: { grault: true },
            waldo: [],
          },
          blah: {},
        };
        value.foo.quux.quuz = value.foo;
        value.foo.garply = value.foo.corge;
        value.foo.waldo.push(value.foo);
        value.blah.huh = value.blah;

        safeJsonStringify(value);
      },
    },
  ],
  { runs: 1000000 }
);
